File: lib/data/models/catalog.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'catalog.freezed.dart';
part 'catalog.g.dart';

/// How cheap this item's cached price is against the same product at every
/// other market nearby (already computed server-side, `api/lib/catalog.js` +
/// `src/lib/catalogScore.js`'s bucketing) — `otimo` / `ok` / `caro`, or
/// `unico` when there's nothing nearby to compare against (fewer than 2
/// markets stocking it).
const catalogBucketLabel = <String, String>{
  'otimo': 'Ótimo preço',
  'ok': 'Preço razoável',
  'caro': 'Tem mais barato perto',
  'unico': 'Preço único encontrado',
};

/// One product cached for a market, `/api/catalog`'s browse item.
@freezed
abstract class CatalogItem with _$CatalogItem {
  const factory CatalogItem({
    String? gtin,
    required String description,
    String? ncm,
    @JsonKey(name: 'priceCents') @Default(0) int priceCents,
    String? fetchedAt,
    @JsonKey(name: 'minCents') int? minCents,
    @JsonKey(name: 'maxCents') int? maxCents,
    @Default(0) int nStores,
    @Default(0) int rank,

    /// Server-computed via `classify()` (same categories as `Categoria`).
    @Default('outros') String category,

    /// `otimo` | `ok` | `caro` | `unico` — see [catalogBucketLabel].
    @Default('unico') String bucket,

    /// How far this price sits between the region's min and max, 0–100.
    /// Null exactly when [bucket] is `unico`.
    int? pct,
  }) = _CatalogItem;

  factory CatalogItem.fromJson(Map<String, dynamic> json) => _$CatalogItemFromJson(json);
}

/// `/api/catalog?market_codigo=X[&category=carnes]` — everything cached for
/// one market, each item scored against the same product at every other
/// nearby market. [categories] is the unfiltered item count per category, so
/// a category filter row can be drawn even while another category is active.
@freezed
abstract class CatalogResponse with _$CatalogResponse {
  const factory CatalogResponse({
    String? marketCodigo,
    @Default(<CatalogItem>[]) List<CatalogItem> items,
    @Default(<String, int>{}) Map<String, int> categories,
  }) = _CatalogResponse;

  factory CatalogResponse.fromJson(Map<String, dynamic> json) => _$CatalogResponseFromJson(json);
}

File: backend/api/catalog.js
import { getMarketCatalog } from './lib/catalog.js';
import { classify } from '../src/lib/categoria.js';
import { cheapnessScore } from '../src/lib/catalogScore.js';
import { applyCors } from './lib/cors.js';

// Browse everything cached for one market (raw_prices, Layer 1), each product scored
// against the same product's price at every other market nearby. GET
// /api/catalog?market_codigo=X[&category=carnes].
export default async function handler(req, res) {
  if (applyCors(req, res)) return;
  const marketCodigo = req.query && req.query.market_codigo;
  if (!marketCodigo) return res.status(400).json({ error: 'missing_market_codigo' });
  const wantCategory = (req.query && req.query.category) || null;
  try {
    const rows = await getMarketCatalog(marketCodigo);
    const items = rows.map((r) => ({
      ...r,
      category: classify({ description: r.description, ncm: r.ncm }),
      ...cheapnessScore(r.priceCents, r.minCents, r.maxCents, r.nStores),
    }));
    const categories = {};
    for (const it of items) categories[it.category] = (categories[it.category] || 0) + 1;
    const filtered = wantCategory ? items.filter((it) => it.category === wantCategory) : items;
    return res.status(200).json({ marketCodigo, items: filtered, categories });
  } catch (e) {
    return res.status(502).json({ error: 'catalog_failed', detail: String(e && e.message) });
  }
}
